### PR TITLE
[WIP] Refactor networking to shared lib

### PR DESCRIPTION
**Pull request by rnystrom**
Testing some stuff out. Liking the direction

Part of #1544 

Interesting bits:
https://github.com/rnystrom/GitHawk/pull/1602/files#diff-9c9d87b8351d92ace94623374d716de8

Tests:
https://github.com/rnystrom/GitHawk/pull/1602/files#diff-6f088fcfc36f6f5c6ca55a34eb89d693

### CODE REVIEW

**Comment by jdisho** [comment]
Is it possible to provide reactive extensions for ReactiveSwift and RxSwift? 😄 @rnystrom

**Comment by BasThomas** [comment]
I am sure that @rnystrom will be open to pull requests once this lib is pulled into its own repository 😊

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by rnystrom** [comment]
Now we're cooking. Added some notifications models and now using `Codable` to unpack the JSON. Much cleaner.

Implementing a network request/response is as simple as this:

```swift
public struct V3NotificationResponse: V3Request {
  // can be single entity of array
  public typealias ResponseType = V3Response<[V3Notification]>
  // other endpoints can init w/ params (owner, repo)
  public let pathComponents: [String] = ["notifications"]

  // other settings defaulted via extension
}
```

Then the `V3Notification` model is just:

```swift
public struct V3Notification: Codable {

    enum Reason: String, Codable {
        case assign = "assign"
        case author = "author"
        case comment = "comment"
        case invitation = "invitation"
        case manual = "manual"
        case mention = "mention"
        case stateChange = "state_change"
        case subscribed = "subscribed"
        case teamMention = "team_mention"
        case reviewRequested = "review_requested"
    }

    let id: String
    let lastReadAt: Date?
    let reason: Reason
    let repository: V3Repository
    let subject: V3NotificationSubject
    let unread: Bool
    let updatedAt: Date

    // omitted custom keys

}
```

File: Local Pods/GitHubAPI/GitHubAPI/V3Notification.swift [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        case author = "author"
+        case comment = "comment"
+        case invitation = "invitation"
+        case manual = "manual"
+        case mention = "mention"
+        case stateChange = "state_change"
+        case subscribed = "subscribed"
+        case teamMention = "team_mention"
+        case reviewRequested = "review_requested"
+    }
```
Comment (jdisho) [inline Local Pods/GitHubAPI/GitHubAPI/V3Notification.swift]: Why not like this?

```
enum Reason: String, Codable {
  case assign
  case author
  case comment
  case invitation
  case manual
  case mention
  case stateChange
  case subscribed
  case teamMention = "team_mention"
  case reviewRequested = "review_requested"
}
```

Comment (rnystrom) [reply Local Pods/GitHubAPI/GitHubAPI/V3Notification.swift]: Used a script to gen these, but ya same thing 😄

File: Local Pods/GitHubAPI/GitHubAPI/V3PullRequestFilesRequest.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+//  V3PullRequestFilesRequest.swift
+//  GitHubAPI
+//
+//  Created by Ryan Nystrom on 3/4/18.
+//  Copyright © 2018 Ryan Nystrom. All rights reserved.
+//
+
+import Foundation
+
+public struct V3PullRequestFilesRequest: V3Request {
```
Comment (BasThomas) [inline Local Pods/GitHubAPI/GitHubAPI/V3PullRequestFilesRequest.swift]: What about namespacing the `V3`?

File: Classes/Issues/Comments/IssueCommentSectionController.swift [hunk 3]
```diff
@@ -181,10 +182,13 @@ IssueCommentDoubleTapDelegate {
...
-        // Actually delete the comment now
-        client.deleteComment(owner: model.owner, repo: model.repo, commentID: number) { [weak self] result in
+        client.client.send(V3DeleteCommentRequest(
+            owner: model.owner,
+            repo: model.repo,
+            commentID: "\(number)")
+        ) { [weak self] result in
             switch result {
-            case .error:
+            case .failure:
```
Comment (SD10) [inline Classes/Issues/Comments/IssueCommentSectionController.swift]: ❤️

File: Local Pods/GitHubAPI/GitHubAPI/V3PullRequestFilesRequest.swift (hunk 2 above)
Comment (rnystrom) [reply Local Pods/GitHubAPI/GitHubAPI/V3PullRequestFilesRequest.swift]: That’d require a separate framework right?

**Comment by rnystrom** [comment]
It’s a big change, but really like how this separates requests into typed structs, and is shareable between libs.

Having the `Client` as part of the `GitHubClient` is only temporary. Ideally we completely remove that.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: Local Pods/GitHubAPI/GitHubAPI/V3PullRequestFilesRequest.swift (hunk 2 above)
Comment (BasThomas) [reply Local Pods/GitHubAPI/GitHubAPI/V3PullRequestFilesRequest.swift]: Oh sorry, I meant having them nested. Like `V3.PullRequestFileRequest`.

Comment (rnystrom) [reply Local Pods/GitHubAPI/GitHubAPI/V3PullRequestFilesRequest.swift]: Ooooh I see! That seems doable. And the benefit is we could have GQL requests if the same name without collisions?

**Pull request merged**